File: mac/Sources/ClipShareCore/ClipMessage.swift
import CoreFoundation
import Foundation

public enum DeviceID: String, Equatable {
    case mac
    case android
}

public enum ClipMessage: Equatable {
    case auth(token: String, deviceId: DeviceID)
    case authOk
    case clip(text: String, deviceId: DeviceID, ts: Int)

    public func encode() -> String {
        let object: [String: Any]

        switch self {
        case let .auth(token, deviceId):
            object = [
                "type": "auth",
                "token": token,
                "deviceId": deviceId.rawValue
            ]
        case .authOk:
            object = ["type": "auth_ok"]
        case let .clip(text, deviceId, ts):
            object = [
                "type": "clip",
                "text": text,
                "deviceId": deviceId.rawValue,
                "ts": ts
            ]
        }

        // Every value above is a JSON-compatible Foundation type.
        let data = try! JSONSerialization.data(
            withJSONObject: object,
            options: [.sortedKeys]
        )
        return String(decoding: data, as: UTF8.self)
    }

    public static func decode(_ json: String) -> ClipMessage? {
        guard let data = json.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data),
              let fields = object as? [String: Any],
              let type = fields["type"] as? String else {
            return nil
        }

        switch type {
        case "auth":
            guard let token = fields["token"] as? String,
                  let deviceIDValue = fields["deviceId"] as? String,
                  let deviceId = DeviceID(rawValue: deviceIDValue) else {
                return nil
            }
            return .auth(token: token, deviceId: deviceId)

        case "auth_ok":
            return .authOk

        case "clip":
            guard let text = fields["text"] as? String,
                  let deviceIDValue = fields["deviceId"] as? String,
                  let deviceId = DeviceID(rawValue: deviceIDValue),
                  let timestamp = fields["ts"] as? NSNumber,
                  CFGetTypeID(timestamp) != CFBooleanGetTypeID(),
                  !CFNumberIsFloatType(timestamp),
                  let ts = timestamp as? Int else {
                return nil
            }
            return .clip(text: text, deviceId: deviceId, ts: ts)

        default:
            return nil
        }
    }
}

File: android/app/src/main/java/com/ymac/clipshare/core/ClipMessage.kt
package com.ymac.clipshare.core

import org.json.JSONException
import org.json.JSONTokener
import org.json.JSONObject

enum class DeviceId(val jsonValue: String) {
    MAC("mac"),
    ANDROID("android");

    companion object {
        fun fromJsonValue(value: String): DeviceId? =
            entries.firstOrNull { it.jsonValue == value }
    }
}

sealed class ClipMessage {
    data class Auth(
        val token: String,
        val deviceId: DeviceId,
    ) : ClipMessage()

    data object AuthOk : ClipMessage()

    data class Clip(
        val text: String,
        val deviceId: DeviceId,
        val ts: Long,
    ) : ClipMessage()

    fun encode(): String {
        val json = JSONObject()

        when (this) {
            is Auth -> json
                .put("type", "auth")
                .put("token", token)
                .put("deviceId", deviceId.jsonValue)

            AuthOk -> json.put("type", "auth_ok")

            is Clip -> json
                .put("type", "clip")
                .put("text", text)
                .put("deviceId", deviceId.jsonValue)
                .put("ts", ts)
        }

        return json.toString()
    }

    companion object {
        fun decode(json: String): ClipMessage? = try {
            val tokener = JSONTokener(json)
            val fields = tokener.nextValue()
            if (fields !is JSONObject || tokener.nextClean() != '\u0000') {
                null
            } else {
                when (val type = fields.requiredString("type")) {
                    "auth" -> decodeAuth(fields)
                    "auth_ok" -> AuthOk
                    "clip" -> decodeClip(fields)
                    else -> null
                }
            }
        } catch (_: JSONException) {
            null
        }

        private fun decodeAuth(fields: JSONObject): Auth? {
            val token = fields.requiredString("token") ?: return null
            val deviceId = fields.requiredDeviceId() ?: return null
            return Auth(token = token, deviceId = deviceId)
        }

        private fun decodeClip(fields: JSONObject): Clip? {
            val text = fields.requiredString("text") ?: return null
            val deviceId = fields.requiredDeviceId() ?: return null
            val ts = fields.requiredLong("ts") ?: return null
            return Clip(text = text, deviceId = deviceId, ts = ts)
        }

        private fun JSONObject.requiredDeviceId(): DeviceId? =
            requiredString("deviceId")?.let(DeviceId::fromJsonValue)

        private fun JSONObject.requiredString(name: String): String? =
            get(name) as? String

        private fun JSONObject.requiredLong(name: String): Long? =
            when (val value = get(name)) {
                is Int -> value.toLong()
                is Long -> value
                else -> null
            }
    }
}
